File: .claude/skills/seedream-5-0-pro/scripts/build-prompt.py
#!/usr/bin/env python3
"""Assemble and validate a Seedream 5.0 Pro prompt for front-end use.

This is a prompt-assembly helper: it builds the natural-language prompt
string a user pastes into a Seedream 5.0 front-end (Dreamina / Jimeng,
Doubao, fal, wavespeed, ComfyUI). It does NOT call any API. On-screen
settings (aspect ratio, resolution, image count) are recorded for the
operator's reference and printed in a trailing settings block; they are
not part of the prompt the model reads.

Modes: T2I, Design (high-density info layout), Edit (grounded precision
edit), MR (reference by ordinal "Image N"), sequential-set.

Seedream 5.0 prompt rules enforced here:
  - subject-first natural-language prose in SPACE order (Subject, Palette
    & style, Arrangement, Camera & light, Extra detail); earlier concepts
    are weighted more heavily.
  - 30-100 word budget; a hard ceiling at 120 words (over-long prompts
    confuse the model).
  - in-image text must be wrapped in double quotes (renders literally in
    any of the 14 supported languages).
  - Edit: exactly one reference (the uploaded image); a marked target and
    an optional recolour value that must be a hex code or a swatch/Image
    reference.
  - MR: 1..10 references, addressed by on-screen order (Image 1, Image 2…),
    identity anchored on Image 1.
  - sequential-set: a named set count (2..8), one global style lock.
  - resolution ceiling is 2K (2048 px) — a 4K request is rejected.
  - layer separation is a cross-cutting request appended to the prompt.
  - no negation-only prompts (Seedream has no negative-prompt field).
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from typing import Any, List, Optional


VALID_MODES = {"T2I", "Design", "Edit", "MR", "sequential-set"}
VALID_ASPECT_RATIOS = {
    "1:1", "16:9", "9:16", "4:5", "5:4", "3:4", "4:3", "3:2", "2:3",
    "21:9", "auto",
}
# Seedream 5.0 Pro is a 2K model: 1K or 2K (max 2048 px), or auto.
# There is deliberately NO 4K tier (this differs from Seedream 4.5).
VALID_RESOLUTIONS = {"1K", "2K", "auto"}
NEGATION_TOKENS = (" no ", " not ", " without ", " avoid ", " never ")
WORD_BUDGET_MIN = 30
WORD_BUDGET_MAX = 100
WORD_HARD_CEILING = 120
MR_REF_CAP = 10
SET_COUNT_MIN = 2
SET_COUNT_MAX = 8
HEX_RE = re.compile(r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")


class PromptValidationError(ValueError):
    pass


def _has_negation(text: str) -> bool:
    if not text:
        return False
    padded = " " + text.lower().strip() + " "
    return any(tok in padded for tok in NEGATION_TOKENS)


def _all_negation(*clauses: str) -> bool:
    populated = [c for c in clauses if c and c.strip()]
    if not populated:
        return False
    return all(_has_negation(c) for c in populated)


def _word_count(*clauses: str) -> int:
    return sum(len(c.split()) for c in clauses if c and c.strip())


def _validate_color(color: str) -> str:
    """A recolour/material value must be a hex code or a swatch/Image
    reference. Returns the normalised value or raises."""
    c = color.strip()
    if HEX_RE.match(c):
        return c
    low = c.lower()
    if "swatch" in low or re.search(r"image\s*\d", low):
        return c
    raise PromptValidationError(
        "recolour value must be a hex code (e.g. '#1E3A8A') or reference a "
        "swatch / an uploaded image (e.g. 'swatch: teal', 'match Image 2'); "
        f"got {color!r}."
    )


def _normalize_refs(raw: List[Any]) -> List[str]:
    out: List[str] = []
    for r in raw:
        if isinstance(r, dict):
            role = r.get("role")
            if not role or not str(role).strip():
                raise PromptValidationError(
                    "every reference requires a non-empty 'role'."
                )
            out.append(str(role))
        else:
            out.append(str(r))
    return out


def _settings_block(*, aspect_ratio: str, resolution: str,
                    set_count: Optional[int]) -> List[str]:
    block = [
        "",
        "[on-screen settings (pick these in the UI; NOT part of the "
        "prompt the model reads):",
        f"  aspect_ratio={aspect_ratio}",
        f"  resolution={resolution}",
    ]
    if set_count is not None:
        block.append(f"  image_count={set_count}")
    block.append("]")
    return block


def _layers_line() -> str:
    return (
        "Layer separation: split the result into independent transparent "
        "layers (title / subject / background / elements); export as PNG."
    )


def build_prompt(
    *,
    mode: str,
    subject: str,
    action: str = "",
    setting: str = "",
    style: str = "",
    camera: str = "",
    references: Optional[List[Any]] = None,
    frames: Optional[List[str]] = None,
    set_count: Optional[int] = None,
    target: Optional[str] = None,
    color: Optional[str] = None,
    layers: bool = False,
    in_image_text: Optional[str] = None,
    text_language: Optional[str] = None,
    aspect_ratio: str = "1:1",
    resolution: str = "2K",
) -> str:
    if mode not in VALID_MODES:
        raise PromptValidationError(
            f"mode must be one of {sorted(VALID_MODES)}; got {mode!r}"
        )
    if aspect_ratio not in VALID_ASPECT_RATIOS:
        raise PromptValidationError(
            f"aspect_ratio must be one of {sorted(VALID_ASPECT_RATIOS)}; "
            f"got {aspect_ratio!r}"
        )
    if resolution not in VALID_RESOLUTIONS:
        raise PromptValidationError(
            f"resolution must be one of {sorted(VALID_RESOLUTIONS)}; got "
            f"{resolution!r}. Seedream 5.0 Pro is a 2K model — there is no "
            "4K tier."
        )
    if in_image_text is not None and in_image_text.strip():
        if '"' not in in_image_text:
            raise PromptValidationError(
                "in-image text must be wrapped in double quotes so Seedream "
                'renders it literally, e.g. \'the title "VISIT KYOTO"\'.'
            )

    if mode == "sequential-set":
        return _build_sequential_set(
            subject=subject,
            style=style,
            frames=frames,
            set_count=set_count,
            layers=layers,
            in_image_text=in_image_text,
            text_language=text_language,
            aspect_ratio=aspect_ratio,
            resolution=resolution,
        )

    return _build_still(
        mode=mode,
        subject=subject,
        action=action,
        setting=setting,
        style=style,
        camera=camera,
        references=references,
        target=target,
        color=color,
        layers=layers,
        in_image_text=in_image_text,
        text_language=text_language,
        aspect_ratio=aspect_ratio,
        resolution=resolution,
    )


def _build_still(
    *,
    mode: str,
    subject: str,
    action: str,
    setting: str,
    style: str,
    camera: str,
    references: Optional[List[Any]],
    target: Optional[str],
    color: Optional[str],
    layers: bool,
    in_image_text: Optional[str],
    text_language: Optional[str],
    aspect_ratio: str,
    resolution: str,
) -> str:
    if not subject or not subject.strip():
        raise PromptValidationError(f"{mode} requires a subject clause.")
    if not style or not style.strip():
        raise PromptValidationError(f"{mode} requires a style clause.")

    if _all_negation(subject, action, setting, style, camera):
        raise PromptValidationError(
            "every clause leans on negation tokens. Rephrase to positive "
            "description; Seedream has no negative-prompt field and does "
            "not honour in-prompt negation reliably."
        )

    refs: List[str] = []
    normalised_color: Optional[str] = None
    if mode == "Edit":
        refs = _normalize_refs(references or [])
        if len(refs) != 1:
            raise PromptValidationError(
                f"Edit mode requires exactly one reference (the uploaded "
                f"image); got {len(refs)}."
            )
        if color is not None and color.strip():
            normalised_color = _validate_color(color)
    elif mode == "MR":
        refs = _normalize_refs(references or [])
        if not (1 <= len(refs) <= MR_REF_CAP):
            raise PromptValidationError(
                f"MR mode requires 1..{MR_REF_CAP} references; got "
                f"{len(refs)}."
            )

    wc = _word_count(subject, action, setting, style, camera)
    if wc > WORD_HARD_CEILING:
        raise PromptValidationError(
            f"prompt is {wc} words; over {WORD_HARD_CEILING} confuses "
            f"Seedream. Trim toward the {WORD_BUDGET_MIN}-{WORD_BUDGET_MAX} "
            "word budget — cut adjectives, not subjects."
        )

    parts: List[str] = []
    if mode == "Design":
        parts.append("High-density design brief:")
    elif mode == "Edit":
        parts.append(f"Reference: {refs[0]}.")
        if target and target.strip():
            parts.append(f"Target: {target.strip()}.")
        parts.append("")
    elif mode == "MR":
        parts.append("Using the uploaded images:")
        for i, role in enumerate(refs, start=1):
            parts.append(f"  Image {i}: {role}.")
        parts.append("")

    for c in (subject, action, setting, style, camera):
        if c and c.strip():
            parts.append(c.strip())

    if normalised_color:
        parts.append(f"Colour / material value: {normalised_color}.")

    if in_image_text and in_image_text.strip():
        line = f"In-image text: {in_image_text.strip()}."
        if text_language and text_language.strip():
            line += f" (language: {text_language.strip()})"
        parts.append(line)

    if layers:
        parts.append(_layers_line())

    parts.extend(_settings_block(
        aspect_ratio=aspect_ratio, resolution=resolution, set_count=None,
    ))
    return "\n".join(parts)


def _build_sequential_set(
    *,
    subject: str,
    style: str,
    frames: Optional[List[str]],
    set_count: Optional[int],
    layers: bool,
    in_image_text: Optional[str],
    text_language: Optional[str],
    aspect_ratio: str,
    resolution: str,
) -> str:
    if not subject or not subject.strip():
        raise PromptValidationError(
            "sequential-set requires a subject held constant across the set."
        )
    if not style or not style.strip():
        raise PromptValidationError(
            "sequential-set requires a single global style/lighting clause "
            "held constant across the whole set."
        )
    if set_count is None:
        raise PromptValidationError(
            "sequential-set requires an explicit set count (e.g. --set-count "
            "6). Without it the model returns a single image."
        )
    if not (SET_COUNT_MIN <= set_count <= SET_COUNT_MAX):
        raise PromptValidationError(
            f"sequential-set count must be {SET_COUNT_MIN}..{SET_COUNT_MAX}; "
            f"got {set_count}. For larger sets, split into two and share an "
            "anchor image to hold identity."
        )
    frames = frames or []
    if frames and len(frames) > set_count:
        raise PromptValidationError(
            f"got {len(frames)} per-image states but the set count is "
            f"{set_count}; states must not exceed the count."
        )

    parts: List[str] = [
        f"Generate a set of {set_count} images of {subject.strip()}.",
        f"Style and lighting (same across the whole set): {style.strip()}.",
    ]
    if frames:
        parts.append("Across the set, change only:")
        for f in frames:
            parts.append(f"  {f.strip()}")
    if in_image_text and in_image_text.strip():
        line = f"In-image text: {in_image_text.strip()}."
        if text_language and text_language.strip():
            line += f" (language: {text_language.strip()})"
        parts.append(line)
    if layers:
        parts.append(_layers_line())

    parts.extend(_settings_block(
        aspect_ratio=aspect_ratio, resolution=resolution,
        set_count=set_count,
    ))
    return "\n".join(parts)


def _parse_refs_arg(raw: Optional[List[str]]) -> Optional[List[Any]]:
    if not raw:
        return None
    out: List[Any] = []
    for r in raw:
        try:
            out.append(json.loads(r))
        except json.JSONDecodeError:
            out.append(r)
    return out


def _parse_args(argv: List[str]) -> argparse.Namespace:
    p = argparse.ArgumentParser(
        description="Assemble a Seedream 5.0 Pro prompt with validation."
    )
    p.add_argument("--mode", required=True, choices=sorted(VALID_MODES))
    p.add_argument("--subject", required=True)
    p.add_argument("--action", default="")
    p.add_argument("--setting", default="")
    p.add_argument("--style", default="")
    p.add_argument("--camera", default="")
    p.add_argument(
        "--reference", action="append", default=None,
        help='Reference role; plain string or JSON {"role": "..."}. '
             "Order = on-screen order (Image 1, Image 2, ...).",
    )
    p.add_argument(
        "--frame", action="append", default=None,
        help="Per-image state for sequential-set mode; repeat per image.",
    )
    p.add_argument("--set-count", type=int, default=None,
                   help="Number of images in a sequential set (2..8).")
    p.add_argument("--target", default=None,
                   help="Edit mode: the marked element (point / box / lasso "
                        "/ sketch), described in words.")
    p.add_argument("--color", default=None,
                   help="Edit mode: recolour / material value — a hex code "
                        "(e.g. '#1E3A8A') or a swatch / Image reference.")
    p.add_argument("--layers", action="store_true",
                   help="Request layer separation (transparent PNG layers).")
    p.add_argument("--in-image-text", default=None,
                   help='Literal text to render; MUST be double-quoted, '
                        'e.g. \'the title "VISIT KYOTO"\'.')
    p.add_argument("--text-language", default=None,
                   help="Optional language hint for in-image text (e.g. ar, "
                        "ja); one of the 14 supported languages.")
    p.add_argument("--aspect-ratio", default="1:1",
                   choices=sorted(VALID_ASPECT_RATIOS))
    p.add_argument("--resolution", default="2K",
                   choices=sorted(VALID_RESOLUTIONS))
    return p.parse_args(argv)


def main(argv: List[str]) -> int:
    args = _parse_args(argv)
    try:
        refs = _parse_refs_arg(args.reference)
        prompt = build_prompt(
            mode=args.mode,
            subject=args.subject,
            action=args.action,
            setting=args.setting,
            style=args.style,
            camera=args.camera,
            references=refs,
            frames=args.frame,
            set_count=args.set_count,
            target=args.target,
            color=args.color,
            layers=args.layers,
            in_image_text=args.in_image_text,
            text_language=args.text_language,
            aspect_ratio=args.aspect_ratio,
            resolution=args.resolution,
        )
    except PromptValidationError as exc:
        print(f"PromptValidationError: {exc}", file=sys.stderr)
        return 2
    print(prompt)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: .claude/skills/seedream-5-0-pro/scripts/build-prompt.ts
#!/usr/bin/env -S npx tsx
/**
 * Assemble and validate a Seedream 5.0 Pro prompt for front-end use.
 *
 * Ported from build-prompt.py (issue #255) — same clause/word-budget/colour-validation rules, same CLI
 * shape and exit codes; only the language changed.
 *
 * This is a prompt-assembly helper: it builds the natural-language prompt string a user pastes into a
 * Seedream 5.0 front-end (Dreamina / Jimeng, Doubao, fal, wavespeed, ComfyUI). It does NOT call any
 * API. On-screen settings (aspect ratio, resolution, image count) are recorded for the operator's
 * reference and printed in a trailing settings block; they are not part of the prompt the model reads.
 *
 * Modes: T2I, Design (high-density info layout), Edit (grounded precision edit), MR (reference by
 * ordinal "Image N"), sequential-set.
 *
 * Seedream 5.0 prompt rules enforced here:
 *   - subject-first natural-language prose in SPACE order (Subject, Palette & style, Arrangement,
 *     Camera & light, Extra detail); earlier concepts are weighted more heavily.
 *   - 30-100 word budget; a hard ceiling at 120 words (over-long prompts confuse the model).
 *   - in-image text must be wrapped in double quotes (renders literally in any of the 14 supported
 *     languages).
 *   - Edit: exactly one reference (the uploaded image); a marked target and an optional recolour value
 *     that must be a hex code or a swatch/Image reference.
 *   - MR: 1..10 references, addressed by on-screen order (Image 1, Image 2...), identity anchored on
 *     Image 1.
 *   - sequential-set: a named set count (2..8), one global style lock.
 *   - resolution ceiling is 2K (2048 px) — a 4K request is rejected.
 *   - layer separation is a cross-cutting request appended to the prompt.
 *   - no negation-only prompts (Seedream has no negative-prompt field).
 */

const VALID_MODES = ["T2I", "Design", "Edit", "MR", "sequential-set"] as const;
type Mode = (typeof VALID_MODES)[number];

const VALID_ASPECT_RATIOS = [
  "1:1", "16:9", "9:16", "4:5", "5:4", "3:4", "4:3", "3:2", "2:3", "21:9", "auto",
] as const;
type AspectRatio = (typeof VALID_ASPECT_RATIOS)[number];

// Seedream 5.0 Pro is a 2K model: 1K or 2K (max 2048 px), or auto. There is deliberately NO 4K tier
// (this differs from Seedream 4.5).
const VALID_RESOLUTIONS = ["1K", "2K", "auto"] as const;
type Resolution = (typeof VALID_RESOLUTIONS)[number];

const NEGATION_TOKENS = [" no ", " not ", " without ", " avoid ", " never "] as const;
const WORD_BUDGET_MIN = 30;
const WORD_BUDGET_MAX = 100;
const WORD_HARD_CEILING = 120;
const MR_REF_CAP = 10;
const SET_COUNT_MIN = 2;
const SET_COUNT_MAX = 8;
const HEX_RE = /^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/;

export class PromptValidationError extends Error {}

function isMode(value: string): value is Mode {
  return (VALID_MODES as readonly string[]).includes(value);
}
function isAspectRatio(value: string): value is AspectRatio {
  return (VALID_ASPECT_RATIOS as readonly string[]).includes(value);
}
function isResolution(value: string): value is Resolution {
  return (VALID_RESOLUTIONS as readonly string[]).includes(value);
}

function hasNegation(text: string): boolean {
  if (!text) return false;
  const padded = " " + text.toLowerCase().trim() + " ";
  return NEGATION_TOKENS.some((tok) => padded.includes(tok));
}

function allNegation(...clauses: string[]): boolean {
  const populated = clauses.filter((c) => c && c.trim());
  if (populated.length === 0) return false;
  return populated.every((c) => hasNegation(c));
}

function wordCount(...clauses: string[]): number {
  return clauses
    .filter((c) => c && c.trim())
    .reduce((total, c) => total + c.split(/\s+/).filter((w) => w.length > 0).length, 0);
}

/** A recolour/material value must be a hex code or a swatch/Image reference. Returns the normalised
 *  value or throws. */
function validateColor(color: string): string {
  const c = color.trim();
  if (HEX_RE.test(c)) return c;
  const low = c.toLowerCase();
  if (low.includes("swatch") || /image\s*\d/.test(low)) return c;
  throw new PromptValidationError(
    "recolour value must be a hex code (e.g. '#1E3A8A') or reference a swatch / an uploaded image " +
      `(e.g. 'swatch: teal', 'match Image 2'); got ${JSON.stringify(color)}.`,
  );
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeRefs(raw: readonly unknown[]): string[] {
  const out: string[] = [];
  for (const r of raw) {
    if (isPlainObject(r)) {
      const role = r["role"];
      if (role === undefined || role === null || String(role).trim() === "") {
        throw new PromptValidationError("every reference requires a non-empty 'role'.");
      }
      out.push(String(role));
    } else {
      out.push(String(r));
    }
  }
  return out;
}

function settingsBlock(aspectRatio: string, resolution: string, setCount: number | undefined): string[] {
  const block = [
    "",
    "[on-screen settings (pick these in the UI; NOT part of the prompt the model reads):",
    `  aspect_ratio=${aspectRatio}`,
    `  resolution=${resolution}`,
  ];
  if (setCount !== undefined) block.push(`  image_count=${setCount}`);
  block.push("]");
  return block;
}

function layersLine(): string {
  return (
    "Layer separation: split the result into independent transparent layers (title / subject / " +
    "background / elements); export as PNG."
  );
}

export interface BuildPromptArgs {
  readonly mode: string;
  readonly subject: string;
  readonly action?: string;
  readonly setting?: string;
  readonly style?: string;
  readonly camera?: string;
  readonly references?: readonly unknown[] | undefined;
  readonly frames?: readonly string[] | undefined;
  readonly setCount?: number | undefined;
  readonly target?: string | undefined;
  readonly color?: string | undefined;
  readonly layers?: boolean;
  readonly inImageText?: string | undefined;
  readonly textLanguage?: string | undefined;
  readonly aspectRatio?: string;
  readonly resolution?: string;
}

export function buildPrompt(args: BuildPromptArgs): string {
  const {
    mode,
    subject,
    action = "",
    setting = "",
    style = "",
    camera = "",
    references,
    frames,
    setCount,
    target,
    color,
    layers = false,
    inImageText,
    textLanguage,
    aspectRatio = "1:1",
    resolution = "2K",
  } = args;

  if (!isMode(mode)) {
    throw new PromptValidationError(
      `mode must be one of ${JSON.stringify([...VALID_MODES].sort())}; got ${JSON.stringify(mode)}`,
    );
  }
  if (!isAspectRatio(aspectRatio)) {
    throw new PromptValidationError(
      `aspect_ratio must be one of ${JSON.stringify([...VALID_ASPECT_RATIOS].sort())}; ` +
        `got ${JSON.stringify(aspectRatio)}`,
    );
  }
  if (!isResolution(resolution)) {
    throw new PromptValidationError(
      `resolution must be one of ${JSON.stringify([...VALID_RESOLUTIONS].sort())}; got ` +
        `${JSON.stringify(resolution)}. Seedream 5.0 Pro is a 2K model — there is no 4K tier.`,
    );
  }
  if (inImageText !== undefined && inImageText.trim()) {
    if (!inImageText.includes('"')) {
      throw new PromptValidationError(
        "in-image text must be wrapped in double quotes so Seedream renders it literally, e.g. " +
          "'the title \"VISIT KYOTO\"'.",
      );
    }
  }

  if (mode === "sequential-set") {
    return buildSequentialSet({
      subject,
      style,
      frames,
      setCount,
      layers,
      inImageText,
      textLanguage,
      aspectRatio,
      resolution,
    });
  }

  return buildStill({
    mode,
    subject,
    action,
    setting,
    style,
    camera,
    references,
    target,
    color,
    layers,
    inImageText,
    textLanguage,
    aspectRatio,
    resolution,
  });
}

function buildStill(opts: {
  mode: Mode;
  subject: string;
  action: string;
  setting: string;
  style: string;
  camera: string;
  references: readonly unknown[] | undefined;
  target: string | undefined;
  color: string | undefined;
  layers: boolean;
  inImageText: string | undefined;
  textLanguage: string | undefined;
  aspectRatio: string;
  resolution: string;
}): string {
  const { mode, subject, action, setting, style, camera, references, target, color, layers, inImageText, textLanguage, aspectRatio, resolution } = opts;

  if (!subject || !subject.trim()) throw new PromptValidationError(`${mode} requires a subject clause.`);
  if (!style || !style.trim()) throw new PromptValidationError(`${mode} requires a style clause.`);

  if (allNegation(subject, action, setting, style, camera)) {
    throw new PromptValidationError(
      "every clause leans on negation tokens. Rephrase to positive description; Seedream has no " +
        "negative-prompt field and does not honour in-prompt negation reliably.",
    );
  }

  let refs: string[] = [];
  let normalisedColor: string | undefined;
  if (mode === "Edit") {
    refs = normalizeRefs(references ?? []);
    if (refs.length !== 1) {
      throw new PromptValidationError(`Edit mode requires exactly one reference (the uploaded image); got ${refs.length}.`);
    }
    if (color !== undefined && color.trim()) {
      normalisedColor = validateColor(color);
    }
  } else if (mode === "MR") {
    refs = normalizeRefs(references ?? []);
    if (!(refs.length >= 1 && refs.length <= MR_REF_CAP)) {
      throw new PromptValidationError(`MR mode requires 1..${MR_REF_CAP} references; got ${refs.length}.`);
    }
  }

  const wc = wordCount(subject, action, setting, style, camera);
  if (wc > WORD_HARD_CEILING) {
    throw new PromptValidationError(
      `prompt is ${wc} words; over ${WORD_HARD_CEILING} confuses Seedream. Trim toward the ` +
        `${WORD_BUDGET_MIN}-${WORD_BUDGET_MAX} word budget — cut adjectives, not subjects.`,
    );
  }

  const parts: string[] = [];
  if (mode === "Design") {
    parts.push("High-density design brief:");
  } else if (mode === "Edit") {
    parts.push(`Reference: ${refs[0]}.`);
    if (target && target.trim()) parts.push(`Target: ${target.trim()}.`);
    parts.push("");
  } else if (mode === "MR") {
    parts.push("Using the uploaded images:");
    refs.forEach((role, i) => parts.push(`  Image ${i + 1}: ${role}.`));
    parts.push("");
  }

  for (const c of [subject, action, setting, style, camera]) {
    if (c && c.trim()) parts.push(c.trim());
  }

  if (normalisedColor) {
    parts.push(`Colour / material value: ${normalisedColor}.`);
  }

  if (inImageText && inImageText.trim()) {
    let line = `In-image text: ${inImageText.trim()}.`;
    if (textLanguage && textLanguage.trim()) line += ` (language: ${textLanguage.trim()})`;
    parts.push(line);
  }

  if (layers) {
    parts.push(layersLine());
  }

  parts.push(...settingsBlock(aspectRatio, resolution, undefined));
  return parts.join("\n");
}

function buildSequentialSet(opts: {
  subject: string;
  style: string;
  frames: readonly string[] | undefined;
  setCount: number | undefined;
  layers: boolean;
  inImageText: string | undefined;
  textLanguage: string | undefined;
  aspectRatio: string;
  resolution: string;
}): string {
  const { subject, style, frames: framesArg, setCount, layers, inImageText, textLanguage, aspectRatio, resolution } = opts;

  if (!subject || !subject.trim()) {
    throw new PromptValidationError("sequential-set requires a subject held constant across the set.");
  }
  if (!style || !style.trim()) {
    throw new PromptValidationError(
      "sequential-set requires a single global style/lighting clause held constant across the whole set.",
    );
  }
  if (setCount === undefined) {
    throw new PromptValidationError(
      "sequential-set requires an explicit set count (e.g. --set-count 6). Without it the model " +
        "returns a single image.",
    );
  }
  if (!(setCount >= SET_COUNT_MIN && setCount <= SET_COUNT_MAX)) {
    throw new PromptValidationError(
      `sequential-set count must be ${SET_COUNT_MIN}..${SET_COUNT_MAX}; got ${setCount}. For larger ` +
        "sets, split into two and share an anchor image to hold identity.",
    );
  }
  const frames = framesArg ?? [];
  if (frames.length > 0 && frames.length > setCount) {
    throw new PromptValidationError(
      `got ${frames.length} per-image states but the set count is ${setCount}; states must not ` +
        "exceed the count.",
    );
  }

  const parts: string[] = [
    `Generate a set of ${setCount} images of ${subject.trim()}.`,
    `Style and lighting (same across the whole set): ${style.trim()}.`,
  ];
  if (frames.length > 0) {
    parts.push("Across the set, change only:");
    for (const f of frames) parts.push(`  ${f.trim()}`);
  }
  if (inImageText && inImageText.trim()) {
    let line = `In-image text: ${inImageText.trim()}.`;
    if (textLanguage && textLanguage.trim()) line += ` (language: ${textLanguage.trim()})`;
    parts.push(line);
  }
  if (layers) {
    parts.push(layersLine());
  }

  parts.push(...settingsBlock(aspectRatio, resolution, setCount));
  return parts.join("\n");
}

function parseRefsArg(raw: readonly string[] | undefined): unknown[] | undefined {
  if (!raw || raw.length === 0) return undefined;
  const out: unknown[] = [];
  for (const r of raw) {
    try {
      out.push(JSON.parse(r));
    } catch {
      out.push(r);
    }
  }
  return out;
}

interface ParsedArgs {
  mode: string;
  subject: string;
  action: string;
  setting: string;
  style: string;
  camera: string;
  reference: string[] | undefined;
  frame: string[] | undefined;
  setCount: number | undefined;
  target: string | undefined;
  color: string | undefined;
  layers: boolean;
  inImageText: string | undefined;
  textLanguage: string | undefined;
  aspectRatio: string;
  resolution: string;
}

const HELP_TEXT = `usage: build-prompt.ts --mode {T2I,Design,Edit,MR,sequential-set} --subject SUBJECT [options]

Assemble a Seedream 5.0 Pro prompt with validation.

options:
  --mode {T2I,Design,Edit,MR,sequential-set}   required
  --subject SUBJECT                              required
  --action ACTION
  --setting SETTING
  --style STYLE
  --camera CAMERA
  --reference REFERENCE                              plain string or JSON {"role": "..."}.
  --frame FRAME                                        per-image state for sequential-set mode.
  --set-count SET_COUNT                                  number of images in a sequential set (2..8).
  --target TARGET                                          Edit mode: the marked element, in words.
  --color COLOR                                              Edit mode: hex code or swatch/Image reference.
  --layers                                                     request layer separation.
  --in-image-text IN_IMAGE_TEXT                                  literal text to render; MUST be double-quoted.
  --text-language TEXT_LANGUAGE                                    optional language hint for in-image text.
  --aspect-ratio {...}   default 1:1
  --resolution {1K,2K,auto}   default 2K
`;

function parseArgs(argv: readonly string[]): ParsedArgs {
  const references: string[] = [];
  let hasReferences = false;
  const frames: string[] = [];
  let hasFrames = false;
  const values: Record<string, string> = {};
  let setCount: number | undefined;
  let layers = false;
  let aspectRatio = "1:1";
  let resolution = "2K";

  let i = 0;
  while (i < argv.length) {
    const token = argv[i]!;
    if (token === "--help" || token === "-h") {
      process.stdout.write(HELP_TEXT);
      process.exit(0);
    }
    if (token === "--layers") {
      layers = true;
      i += 1;
      continue;
    }
    const eq = token.indexOf("=");
    const flag = eq === -1 ? token : token.slice(0, eq);
    const inlineValue = eq === -1 ? undefined : token.slice(eq + 1);
    const next = (): string => {
      if (inlineValue !== undefined) return inlineValue;
      i += 1;
      const v = argv[i];
      if (v === undefined) throw new PromptValidationError(`argument ${flag}: expected one argument`);
      return v;
    };

    switch (flag) {
      case "--mode":
        values["mode"] = next();
        break;
      case "--subject":
        values["subject"] = next();
        break;
      case "--action":
        values["action"] = next();
        break;
      case "--setting":
        values["setting"] = next();
        break;
      case "--style":
        values["style"] = next();
        break;
      case "--camera":
        values["camera"] = next();
        break;
      case "--reference":
        references.push(next());
        hasReferences = true;
        break;
      case "--frame":
        frames.push(next());
        hasFrames = true;
        break;
      case "--set-count":
        setCount = Number.parseInt(next(), 10);
        break;
      case "--target":
        values["target"] = next();
        break;
      case "--color":
        values["color"] = next();
        break;
      case "--in-image-text":
        values["inImageText"] = next();
        break;
      case "--text-language":
        values["textLanguage"] = next();
        break;
      case "--aspect-ratio":
        aspectRatio = next();
        break;
      case "--resolution":
        resolution = next();
        break;
      default:
        throw new PromptValidationError(`unrecognized arguments: ${token}`);
    }
    i += 1;
  }

  if (values["mode"] === undefined) throw new PromptValidationError("the following arguments are required: --mode");
  if (values["subject"] === undefined) throw new PromptValidationError("the following arguments are required: --subject");

  return {
    mode: values["mode"],
    subject: values["subject"],
    action: values["action"] ?? "",
    setting: values["setting"] ?? "",
    style: values["style"] ?? "",
    camera: values["camera"] ?? "",
    reference: hasReferences ? references : undefined,
    frame: hasFrames ? frames : undefined,
    setCount,
    target: values["target"],
    color: values["color"],
    layers,
    inImageText: values["inImageText"],
    textLanguage: values["textLanguage"],
    aspectRatio,
    resolution,
  };
}

function main(argv: readonly string[]): number {
  let args: ParsedArgs;
  try {
    args = parseArgs(argv);
  } catch (exc) {
    if (exc instanceof PromptValidationError) {
      process.stderr.write(`PromptValidationError: ${exc.message}\n`);
      return 2;
    }
    throw exc;
  }

  try {
    const refs = parseRefsArg(args.reference);
    const prompt = buildPrompt({
      mode: args.mode,
      subject: args.subject,
      action: args.action,
      setting: args.setting,
      style: args.style,
      camera: args.camera,
      references: refs,
      frames: args.frame,
      setCount: args.setCount,
      target: args.target,
      color: args.color,
      layers: args.layers,
      inImageText: args.inImageText,
      textLanguage: args.textLanguage,
      aspectRatio: args.aspectRatio,
      resolution: args.resolution,
    });
    process.stdout.write(prompt + "\n");
    return 0;
  } catch (exc) {
    if (exc instanceof PromptValidationError) {
      process.stderr.write(`PromptValidationError: ${exc.message}\n`);
      return 2;
    }
    throw exc;
  }
}

const isMainModule = process.argv[1] !== undefined && import.meta.url === `file://${process.argv[1]}`;
if (isMainModule) {
  process.exitCode = main(process.argv.slice(2));
}
